src/plugins/discourse/mirror.test.js

```diff
@@ -568,4 +654,548 @@ describe("plugins/discourse/mirror", () => {
+        fetcher._post(p2),
+        fetcher._post(p3),
+        fetcher._post(p4),
+      ];
+      await mirror.update(reporter);
+      const actualBefore = repo.posts();
+
+      // Category definition post we don't expect to be updated.
+      fetcher.editPost(p2, {
+        cooked: `<p>Edit: changed category definition post.</p>`,
+        timestampMs: Date.now(),
+      });
+      fetcher.editPost(p4, {
+        cooked: `<p>Edit: changed normal post.</p>`,
+        timestampMs: Date.now(),
+      });
+      const expectedAfter = [
+        fetcher._post(p1),
+        expectedBefore[1],
+        fetcher._post(p3),
+        fetcher._post(p4),
+      ];
+      await mirror.update(reporter);
+      const actualAfter = repo.posts();
+
+      expect(actualBefore).toEqual(expectedBefore);
+      expect(actualAfter).toEqual(expectedAfter);
+      // Sanity check, make sure the category post was edited in the mock fetcher.
+      expect(fetcher._post(p2)).not.toEqual(expectedAfter[1]);
+    });
+
+    it("provides usernames for all active users", async () => {
+      const {mirror, fetcher, reporter, repo} = example();
+      fetcher.addPost(2, null, "alpha");
+      fetcher.addPost(3, null, "beta");
+      fetcher.addPost(3, 1, "alpha");
+      await mirror.update(reporter);
+      // credbot appears because it is the nominal author of all topics
+      expect(
+        repo
+          .users()
+          .slice()
+          .sort()
+      ).toEqual(["alpha", "beta", "credbot"]);
+    });
+
+    function expectLikesSorted(as, bs) {
+      const s = (ls) =>
+        sortBy(
+          ls,
+          (x) => x.username,
+          (x) => x.postId
+        );
+      expect(s(as)).toEqual(s(bs));
+    }
+
+    it("provides all the likes by users that have posted", async () => {
+      const {mirror, fetcher, reporter, repo} = example();
+
+      fetcher.addPost(1, null, "alpha");
+      fetcher.addPost(2, null, "alpha");
+      fetcher.addPost(3, null, "beta");
+      const l1 = fetcher.addLike("beta", 1, 5);
+      const l2 = fetcher.addLike("beta", 2, 6);
+      const l3 = fetcher.addLike("beta", 3, 7);
+      const l4 = fetcher.addLike("alpha", 1, 8);
+      await mirror.update(reporter);
+      expectLikesSorted(repo.likes(), [l1, l2, l3, l4]);
+
+      const l5 = fetcher.addLike("alpha", 2, 9);
+      fetcher.addPost(4, null, "credbot");
+      const l6 = fetcher.addLike("credbot", 2, 10);
+      const l7 = fetcher.addLike("beta", 4, 11);
+      await mirror.update(reporter);
+      expectLikesSorted(repo.likes(), [l1, l2, l3, l4, l5, l6, l7]);
+    });
+
+    it("doesn't find likes of users that never posted", async () => {
+      const {mirror, fetcher, reporter, repo} = example();
+      fetcher.addPost(1, null);
+      fetcher.addLike("nope", 1, 1);
+      await mirror.update(reporter);
+      expect(repo.likes()).toEqual([]);
+    });
+
+    it("mirrors category definition topics only after it's interval", async () => {
```

Comment:
this test description makes me think that the test verifies that the updated category definitions will actually get discovered (after enough time has passed), but that doesn't happen here.

Reply:
maybe we should mock Date.now() so we can fully test the interval logic

Reply:
We could mock time to test that. I've done so before in another project that had several critically important functions that were time related. Such as requiring different behavior depending on the hour-of-day, timezone, date, day-of-week, etc.

Mocking this correctly can be tricky though. From that, I would recommend not mocking Date.now().
But instead adding a dependency to the Mirror constructor, **and all if it's dependencies** from which to source the current time. For example a `getDate: () => Date` function. Forcing ourselves to use that dependency as the source of time in reviews. Basically all across the code base. Reason being, there are other ways to get time that we might not have mocked. `new Date()`, `process.hrtime()`, timestamps created by the database and probably more. Not doing something this rigorously can lead to flaky tests due to the mock only covering part of the time sources, causing unexpected false-positives and false-negatives.

Additionally, bringing in a library like [luxon](https://www.npmjs.com/package/luxon) to make those tests legible, instead of having millisecond time math all over the place.

Right now, I'm dodging having to do that by making use of the fact that 0 ms is a valid interval and it will be executed on the next update even if the time between update calls is <1 ms, while the default test value of `recheckCategoryDefinitionsAfterMs: 3600000, // 1h` is safe to assume won't elapse in a single test case. Although we could use values up to `+Infinity` as well to guarantee this.

The only drawback being, we're not doing automated testing with likely real-world scenarios. Like 2 days between updates or 40 minutes between updates. Which seems like a smoke test to me. Not sure it's worth refactoring for that right now.